File: Java/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        // Ensure nums1 is the smaller array
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.length;
        int n = nums2.length;
        int total = m + n;
        int half = (total + 1) / 2;

        int left = 0, right = m;

        while (left <= right) {
            int i = (left + right) / 2;
            int j = half - i;

            int A_left  = (i == 0) ? Integer.MIN_VALUE : nums1[i - 1];
            int A_right = (i == m) ? Integer.MAX_VALUE : nums1[i];
            int B_left  = (j == 0) ? Integer.MIN_VALUE : nums2[j - 1];
            int B_right = (j == n) ? Integer.MAX_VALUE : nums2[j];

            if (A_left <= B_right && B_left <= A_right) {
                if (total % 2 == 0) {
                    return (Math.max(A_left, B_left) + Math.min(A_right, B_right)) / 2.0;
                } else {
                    return Math.max(A_left, B_left);
                }
            } else if (A_left > B_right) {
                right = i - 1;
            } else {
                left = i + 1;
            }
        }

        // Should never reach here for valid input
        return 0.0;
    }
}

File: Go/0004-median-of-two-sorted-arrays.go
import (
	"math"
)

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	// Ensure nums1 is the smaller array
	if len(nums1) > len(nums2) {
		return findMedianSortedArrays(nums2, nums1)
	}

	m, n := len(nums1), len(nums2)
	total := m + n
	half := (total + 1) / 2

	left, right := 0, m

	for left <= right {
		i := (left + right) / 2
		j := half - i

		A_left := math.Inf(-1)
		if i > 0 {
			A_left = float64(nums1[i-1])
		}
		A_right := math.Inf(1)
		if i < m {
			A_right = float64(nums1[i])
		}

		B_left := math.Inf(-1)
		if j > 0 {
			B_left = float64(nums2[j-1])
		}
		B_right := math.Inf(1)
		if j < n {
			B_right = float64(nums2[j])
		}

		if A_left <= B_right && B_left <= A_right {
			if total%2 == 0 {
				return (math.Max(A_left, B_left) + math.Min(A_right, B_right)) / 2.0
			} else {
				return math.Max(A_left, B_left)
			}
		} else if A_left > B_right {
			right = i - 1
		} else {
			left = i + 1
		}
	}

	// Should not reach here
	return 0.0
}
